[Type checker] Fold more for-each type checking into the constraint solver

DougGregor opened the pull request:
The type checking of the for-each loop is split between the constraint
solver (which does most of the work) and the statement checker (which
updates the for-each loop AST). Move more of the work into the constraint
solver proper, so that the AST updates can happen in one place, making use
of the solution produced by the solver. This allows a few things, some of
which are short-term gains and others that are more future-facing:

* `TypeChecker::convertToType` has been removed, because we can now either
use the more general `typeCheckExpression` entry point or perform the
appropriate operation within the constraint system.
* Solving the constraint system ensures that everything related to the
for-each loop fulyl checks out
* Additional refactoring will make it easier for the for-each loop to be
checked as part of a larger constraint system, e.g., for processing entire
closures or function bodies (that’s the futurist bit).

Use this to get some diagnostic improvements, specifically an improved
“did you mean to unwrap an optional” diagnostic. If you try to loop
over an `[Int]?`, the diagnostic will now be:

    error: for-in loop over optional type '[Int]?' must be unwrapped to a value of type '[Int]'

with the normal set of notes with Fix-Its to suggest reasonable actions:

    note: coalesce using '??' to provide a default when the optional value contains 'nil'
    note: force-unwrap using '!' to abort execution if the optional value contains 'nil'

It also introduces the notion of a "value witness" constraint, which better describes the
operation of looking for a value witness within a particular conformance. If the
conformance itself is ill-formed, then the right-hand-side of the constraint (the member
type) can be filled in with holes to suppress further diagnostics. This also ensures that the
AST records `makeIterator` calls through conformance rather than devirtualizing in the type
checker.

DougGregor committed:
[Type checker] Fold more for-each type checking into the constraint solver.

The type checking of the for-each loop is split between the constraint
solver (which does most of the work) and the statement checker (which
updates the for-each loop AST). Move more of the work into the constraint
solver proper, so that the AST updates can happen in one place, making use
of the solution produced by the solver. This allows a few things, some of
which are short-term gains and others that are more future-facing:

* `TypeChecker::convertToType` has been removed, because we can now either
use the more general `typeCheckExpression` entry point or perform the
appropriate operation within the constraint system.
* Solving the constraint system ensures that everything related to the
for-each loop full checks out
* Additional refactoring will make it easier for the for-each loop to be
checked as part of a larger constraint system, e.g., for processing entire
closures or function bodies (that’s the futurist bit).

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@xedin the last commit is the one you won't be thrilled with; see the `ForEachAuxiliary` locator part. I'll think more about dependent constraint propagation for this case, but at the moment it feels like it would require a lot of machinery, and I'd like to get a better sense of the payoff.

xedin commented on lib/Sema/ConstraintLocator.cpp:
```diff
@@ -473,6 +475,10 @@ void ConstraintLocator::dump(SourceManager *sm, raw_ostream &out) const {
       out << "condition expression";
       break;
+
+    case ForEachAuxiliary:
```
I'd be very happy if we didn't do it this way (which effectively hides the problem) and instead addressed a root case limitation of the constraint system. I feel like this is yet another case when diagnostics are put on the back-burner so somebody else would have deal with that later. I'd like to propose an alternative approach.

The problem here is related to the fact that there is no way to express dependence of `makeIterator` on its base type conforming to `Sequence`. There is, maybe not the most elegant way, to fix that limitation in the current framework by changing how `conforms to` constraint behaves or introducing a new similar constraint which would eliminate a need for special locator because it would be possible to record a non-confirming type as a constraint system "hole". 

`Conforms To` constraint could be extended to support a third type which gets bound when conformance is "confirmed" as a result of successful constraint simplification, this could also be expressed as a separate `Conforms To (Bind)` constraint, such would be used to establish ordering between conformance constraint and its dependents e.g. `T0 <conforms to and bind> Sequence -> T1`, resulting `T1` could be used as a base type for member lookup e.g. `T1[.makeIterator] == T2`. `T1` would only be bound if `T0` conforms to `Sequence` which means that `.makeIterator` constraint cannot be simplified until conformance is checked (because its base type wouldn't be resolved). This also gives the solver an opportunity to mark `T1` as a "hole" if `T0` doesn't conform and just drop `.makeIterator` constraint when needed without having to produce any fixes.

DougGregor replied:
I've reimplemented the last commit with the new "value witness" constraint that does essentially what you describe.

DougGregor committed:
[Type checker] Introduce value witness constraints.

Introduce a new kind of constraint, the "value witness" constraint,
which captures a reference to a witness for a specific protocol
conformance. It otherwise acts like a more restricted form of a "value
member" constraint, where the specific member is known (as a
ValueDecl*) in advance.

The constraint is effectively dependent on the protocol
conformance itself; if that conformance fails, mark the type variables
in the resolved member type as "holes", so that the conformance
failure does not cascade.

Note that the resolved overload for this constraint always refers to
the requirement, rather than the witness, so we will end up recording
witness-method references in the AST rather than concrete references,
and leave it up to the optimizers to perform devirtualization. This is
demonstrated by the SIL changes needed in tests, and is part of the
wider resilience issue with conformances described by
rdar://problem/22708391.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please benchmark
